session_generation.go

```diff
@@ -0,0 +1,102 @@
+package main
+
+import (
+	"errors"
+	"os"
+	"time"
+
+	"github.com/miquella/vaulted/lib"
+)
+
+var (
+	ErrNoSessionIncompatibleWithAssume  = errors.New("--assume generates session credentials, it cannot be combined with --no-session")
+	ErrNoSessionIncompatibleWithRefresh = errors.New("--refresh refreshes session credentials, it cannot be combined with --no-session")
+	ErrNoSessionRequiresVaultName       = errors.New("A vault name must be specified when using --no-session")
+)
+
+type SessionOptions struct {
+	VaultName string
+
+	NoSession bool
+
+	Refresh bool
+	Role    string
+}
+
+func DefaultSession() *vaulted.Session {
+	return &vaulted.Session{
+		SessionVersion: vaulted.SessionVersion,
+
+		Name:       os.Getenv("VAULTED_ENV"),
+		Expiration: time.Now().Add(time.Hour).Truncate(time.Second),
+	}
+}
+
+func GetSessionWithOptions(store vaulted.Store, options *SessionOptions) (*vaulted.Session, error) {
+	// Disabled session credentials
+	if options.NoSession {
+		if options.VaultName == "" {
+			return nil, ErrNoSessionRequiresVaultName
+		} else if options.Refresh {
+			return nil, ErrNoSessionIncompatibleWithRefresh
+		} else if options.Role != "" {
+			return nil, ErrNoSessionIncompatibleWithAssume
+		}
+
+		return getVaultSessionWithNoSession(store, options)
+	}
+
+	var err error
+	var session *vaulted.Session
+
+	// Get a session
+	if options.VaultName == "" {
+		session = DefaultSession()
+	} else {
+		session, err = getVaultSession(store, options)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Assume any role specified
+	if options.Role != "" {
+		return session.AssumeRole(options.Role)
+	}
+
+	return session, nil
+}
+
+func getVaultSessionWithNoSession(store vaulted.Store, options *SessionOptions) (*vaulted.Session, error) {
+	vault, _, err := store.OpenVault(options.VaultName)
+	if err != nil {
+		return nil, err
+	}
+
+	// Change the in-memory vault to forgo temp cred generation
+	if vault.AWSKey != nil {
+		vault.AWSKey.ForgoTempCredGeneration = true
+	}
+
+	// Skip assuming the vault's role
+
```

Comment:
nit: extra space

Reply:
Although this was intentional, I see the argument for not having the newline separating the comment from the line below it.

When I implemented it, I saw assuming the vault's role as a discrete step that was being omitted, so I separated that (skipped) step from the others. Reading back through it now, I don't know that having it separated adds a lot, but I'm inclined to leave it just the same.